source/postprocess/adjoint_kernels.cc

```diff
@@ -0,0 +1,203 @@
+/*
+  Copyright (C) 2021 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file doc/COPYING.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+#include <aspect/simulator.h>
+#include <aspect/postprocess/adjoint_kernels.h>
+#include <aspect/utilities.h>
+
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/fe/fe_values.h>
+
+namespace aspect
+{
+  namespace Postprocess
+  {
+
+
+    template <int dim>
+    std::pair<std::string,std::string>
+    AdjointKernels<dim>::execute (TableHandler &)
+    {
+      if (this->n_compositional_fields() == 0)
+        return std::pair<std::string,std::string>();
+
+      // create a quadrature formula based on the compositional element alone.
+      // be defensive about determining that a compositional field actually exists
+      AssertThrow (this->introspection().base_elements.compositional_fields
+                   != numbers::invalid_unsigned_int,
+                   ExcMessage("This postprocessor cannot be used without compositional fields."));
```

Comment:
It seems like I repeated this assert statement in different files so thought it would be better to check in parameters.cc that if the adjoint stokes solver scheme was chosen that it needs at least 2 compositional fields and that two of them have to have the specific names. Then in this file and in the material model I simply check that they are only called if the solver scheme adjoint stokes is chosen and I also left a note that this then already checks for compositional fields. Let me know if that makes sense or should be changed.